File: src/main/java/xyz/upperlevel/hgame/world/scheduler/Scheduler.java
package xyz.upperlevel.hgame.world.scheduler;

import java.util.HashMap;
import java.util.Map;

public class Scheduler {
    private static final Map<Integer, TaskHandle> tasksById = new HashMap<>();
    private static int id = 0;

    /**
     * Starts a new task.
     *
     * @param delay     the delay in milliseconds.
     * @param repeating may this task be repeated after it ended?
     * @return the id of the created task.
     */
    public static int start(Runnable action, long delay, boolean repeating) {
        tasksById.put(id, new TaskHandle(id, action, delay, repeating));
        return id++;
    }

    /**
     * Starts a new delay task.
     *
     * @param delay the delay in milliseconds.
     * @return the id of the created task.
     */
    public static int start(Runnable action, long delay) {
        return start(action, delay, false); // by default not repeating
    }

    public static void update() {
        long currentTime = System.currentTimeMillis();
        Map<Integer, TaskHandle> tasksByIdCopy = new HashMap<>(tasksById);
        for (TaskHandle handle : tasksByIdCopy.values()) {
            if (handle.isReady(currentTime)) {
                handle.run();
                if (!handle.isRepeating()) {
                    tasksById.remove(handle.getId());
                }
            }
        }
    }

    public static void cancel(int taskId) {
        tasksById.remove(taskId);
    }

    private Scheduler() {
    }
}

File: game/src/main/kotlin/xyz/upperlevel/hgame/world/scheduler/Scheduler.kt
package xyz.upperlevel.hgame.world.scheduler

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

object Scheduler {
    private val tasksById = ConcurrentHashMap<Int, TaskHandle>()
    private val id = AtomicInteger(0)

    /**
     * Starts a new task.
     *
     * @param delay     the delay in milliseconds.
     * @param repeating may this task be repeated after it ended?
     * @return the id of the created task.
     */
    fun start(action: () -> Unit, delay: Long, repeating: Boolean = false): Int {
        val taskId = id.getAndIncrement()
        tasksById[taskId] = TaskHandle(taskId, action, delay, repeating)
        return taskId
    }

    /**
     * Starts a new task.
     *
     * @param delay     the delay in milliseconds.
     * @param repeating may this task be repeated after it ended?
     * @return the id of the created task.
     */
    fun start(action: Runnable, delay: Long, repeating: Boolean = false): Int {
        return start(action::run, delay, repeating)
    }

    fun update() {
        val currentTime = System.currentTimeMillis()
        for (handle in tasksById.values) {
            if (handle.isReady(currentTime)) {
                handle.run()
                if (!handle.isRepeating) {
                    tasksById.remove(handle.id)
                }
            }
        }
    }

    fun cancel(taskId: Int) {
        tasksById.remove(taskId)
    }
}
